File: ccdeskd/internal/protocol/protocol.go
// Package protocol defines the JSON frame types for the ccdesk WebSocket protocol.
package protocol

// Frame types exchanged between client and server.
const (
	TypeAuth     = "auth"
	TypeAttach   = "attach"
	TypeReady    = "ready"
	TypeData     = "data"
	TypeResize   = "resize"
	TypeSessions = "sessions"
	TypePing     = "ping"
	TypePong     = "pong"
	TypeExit     = "exit"
	TypeError    = "error"
)

// Frame is the envelope for all WebSocket messages.
type Frame struct {
	Type string `json:"type"`
}

// AuthFrame is sent by the client as the first message.
type AuthFrame struct {
	Type  string `json:"type"`
	Token string `json:"token"`
}

// AttachFrame requests opening or resuming a session.
type AttachFrame struct {
	Type      string `json:"type"`
	SessionID string `json:"sessionId,omitempty"` // empty = create new
	Cols      uint16 `json:"cols"`
	Rows      uint16 `json:"rows"`
	Workdir   string `json:"workdir,omitempty"` // working directory for new sessions
}

// ReadyFrame confirms attach success.
type ReadyFrame struct {
	Type      string `json:"type"`
	SessionID string `json:"sessionId"`
	Workdir   string `json:"workdir"`
}

// DataFrame carries PTY bytes (base64-encoded).
type DataFrame struct {
	Type    string `json:"type"`
	Payload string `json:"payload"` // base64
}

// ResizeFrame updates remote PTY dimensions.
type ResizeFrame struct {
	Type string `json:"type"`
	Cols uint16 `json:"cols"`
	Rows uint16 `json:"rows"`
}

// SessionInfo describes a single session in the list.
type SessionInfo struct {
	ID      string `json:"id"`
	Title   string `json:"title"`
	Workdir string `json:"workdir"`
	Created string `json:"created"`
}

// SessionsFrame lists all sessions on this machine.
type SessionsFrame struct {
	Type string        `json:"type"`
	List []SessionInfo `json:"list"`
}

// ExitFrame signals that the session process exited.
type ExitFrame struct {
	Type string `json:"type"`
	Code int    `json:"code"`
}

// ErrorFrame reports an error to the client.
type ErrorFrame struct {
	Type    string `json:"type"`
	Message string `json:"message"`
}

File: desktop/src/shared/protocol.ts
// Protocol frame types — mirrors vibe-remoted/internal/protocol/protocol.go

export const FrameType = {
  Auth: 'auth',
  Attach: 'attach',
  Ready: 'ready',
  Data: 'data',
  Resize: 'resize',
  Sessions: 'sessions',
  Ping: 'ping',
  Pong: 'pong',
  Exit: 'exit',
  Error: 'error',
  Notify: 'notify',
} as const;

export type FrameTypeValue = (typeof FrameType)[keyof typeof FrameType];

// --- Client → Server ---

export interface AuthFrame {
  type: typeof FrameType.Auth;
  token: string;
}

export interface AttachFrame {
  type: typeof FrameType.Attach;
  sessionId?: string; // empty = create new
  cols: number;
  rows: number;
  workdir?: string; // working directory for new sessions
}

export interface DataFrameC2S {
  type: typeof FrameType.Data;
  payload: string; // base64
}

export interface ResizeFrame {
  type: typeof FrameType.Resize;
  cols: number;
  rows: number;
}

export interface PingFrame {
  type: typeof FrameType.Ping;
}

// --- Server → Client ---

export interface ReadyFrame {
  type: typeof FrameType.Ready;
  sessionId: string;
  workdir: string;
}

export interface DataFrameS2C {
  type: typeof FrameType.Data;
  payload: string; // base64
}

export interface SessionInfo {
  id: string;
  title: string;
  workdir: string;
  created: string;
}

export interface SessionsFrame {
  type: typeof FrameType.Sessions;
  list: SessionInfo[];
}

export interface ExitFrame {
  type: typeof FrameType.Exit;
  code: number;
}

export interface ErrorFrame {
  type: typeof FrameType.Error;
  message: string;
}

export interface PongFrame {
  type: typeof FrameType.Pong;
}

export interface NotifyFrame {
  type: typeof FrameType.Notify;
  sessionId: string;
  kind: string; // 'idle' | 'waiting' | future kinds (open enum)
  message?: string;
}

// Union of all server→client frames
export type ServerFrame =
  | ReadyFrame
  | DataFrameS2C
  | SessionsFrame
  | ExitFrame
  | ErrorFrame
  | PongFrame
  | NotifyFrame;

// Union of all client→server frames
export type ClientFrame =
  | AuthFrame
  | AttachFrame
  | DataFrameC2S
  | ResizeFrame
  | PingFrame;

// --- Machine config (client-side) ---

export interface MachineConfig {
  name: string;
  addr: string; // tailscale IP or MagicDNS name
  port: number;
  token: string;
}
